In [1]:
import pandas
import pandas as pd

from GA import GA
import optproblems.cec2005 as optproblems

# Test Functions
- F2: Shifted Schwefel’s Problem 1.2
    - Uni-modal without noise
- F4: Shifted Schwefel’s Problem 1.2 with Noise in Fitness
    - Uni-modal with noise
- F8: Shifted Rotated Ackley’s Function with Global Optimum on Bounds
    - Single function
    - Global optimum on bounds
    - Multi-modal
- F13: Shifted Expanded Griewank’s plus Rosenbrock’s Function (F8F2)
    - Expanded Function
    - Multi-modal
- F17: Rotated Version of Hybrid Composition Function with Noise in Fitness
    - Gaussian noise in fitness
    - Multi-modal
    - Huge number of local optima
    - In-deterministic
- F24: Rotated Hybrid Composition Function
    - Composite function
    - Huge number of local optima
    - Multi-modal

# Experiment with GA with Different Hyper Parameters
First we explore how GA evaluates with a set of configurations using only lower numbers of generations

Hyper parameters

In [2]:
# F2 :   Unimodal without noise
# F4 :   Unimodal with noise
# F8 :   Single Function, Global Optimum on Bounds, Multi-modal
# F13:   Expanded Function, Multi-modal
# F17:   Gaussian noise in Fitness, Multi-modal, Huge number of local optima, indeterministic
# F24:   Composite Function, Huge number of local optima, Multi-modal

benchmarks_and_bounds = [optproblems.F2, [-100,100], "F2", optproblems.F4, [-100,100], "F4", optproblems.F8, [-32,32], "F8", optproblems.F13, [-3,1], "F13", optproblems.F17, [-5,5], "F17", optproblems.F24, [-5,5], "F24"]
no_dimensions = [10, 30, 50]
pop_sizes = [5, 10, 15]
num_generations = [50, 150]
tournament_t = [2, 4]
crossover_percentage = [0.5, 0.95, 1]
mutation_rates = [0.2, 0.8, 1]
decreasing_mutation_rate = [True, False]
elites = [0, 1]

average_over_range = 5

Run configuration suite, if that was done than just read the pickle file.

In [3]:
run_configuration_suite = False
if run_configuration_suite:

    list_for_df = []
    for i in range(0, len(benchmarks_and_bounds), 3):
        number_of_functions = (len(benchmarks_and_bounds)/3)
        print(f"Progress: {i/3} / {number_of_functions}")
        bound = benchmarks_and_bounds[i+1]
        benchmark_name = benchmarks_and_bounds[i+2]
        for no_dimension in no_dimensions:
            print(f"\tDimension: {no_dimension}")
            benchmark = benchmarks_and_bounds[i](no_dimension)
            optimal_solution = benchmark.get_optimal_solutions()
            benchmark.evaluate(optimal_solution[0])
            optimal_minimum = optimal_solution[0].objective_values
            for pop_size in pop_sizes:
                print(f"\t\tPopulation: {pop_size}")
                for t in tournament_t:
                    for crossover in crossover_percentage:
                        for mutation in mutation_rates:
                            for dmr in decreasing_mutation_rate:
                                for elite in elites:
                                    for generations in num_generations:
                                        average_fitness = 0
                                        average_difference_from_optimal = 0

                                        best_fitness = 0
                                        best_time = 0
                                        best_difference_from_optimal = 0

                                        for j in range(average_over_range):
                                            algorithm = GA(benchmark, no_dimension, bound, pop_size, num_generations=generations, t=t, crossover_percentage=crossover, mutation_rate=mutation, decreasing_mutation_rate=dmr, elite=elite)
                                            result, time = algorithm.run()
                                            fitness = result.fitness
                                            diff_from_optimal = fitness - optimal_minimum

                                            average_fitness += 1/average_over_range * fitness
                                            average_difference_from_optimal += 1/average_over_range * diff_from_optimal

                                            if j == 0:
                                                best_fitness = fitness
                                                best_difference_from_optimal = diff_from_optimal
                                                best_time = time
                                                continue

                                            if fitness < best_fitness:
                                                best_fitness = fitness

                                            if diff_from_optimal < best_difference_from_optimal:
                                                best_difference_from_optimal = diff_from_optimal

                                            if time < best_time:
                                                best_time = time

                                        result_and_configuration = [benchmark_name, optimal_minimum, best_fitness, best_difference_from_optimal, average_fitness, average_difference_from_optimal, best_time, generations, no_dimension, pop_size, t, crossover, mutation, dmr, elite, bound, average_over_range]
                                        list_for_df.append(result_and_configuration)

    labels = ["TestFunction", "OptimalMinimum", "BestFitness", "BestDiffFromOpt", "AverageFitness", "AverageDiffFromOpt", "TimePerRun", "Generations", "NoDimension", "PopSize", "TForTournament", "CrossoverPerc", "MutationRate", "DecreasingMutationRate", "Elite", "Bounds", "AverageOverRange"]
    df = pandas.DataFrame(data=list_for_df, columns=labels)
    df.to_pickle("GA_df.pkl")
else:
    df = pd.read_pickle("GA_df.pkl")

# F2:

- Best Diff: 312.11 -> 150 Generation / 10 Dimension / 15 Pop Size / 2 T / 100% CR / 20% MR / DMR True / 1 Elite
- Avr. Diff: 913.20 -> 150 Generation / 10 Dimension / 15 Pop Size / 4 T / 95%  CR / 20% MR / DMR True / 1 Elite


In [4]:
df.query("TestFunction == 'F2'")

TestFunction  OptimalMinimum    BestFitness  BestDiffFromOpt  \
0              F2          -450.0   17690.525955     18140.525955   
1              F2          -450.0   14235.020676     14685.020676   
2              F2          -450.0    6013.072511      6463.072511   
3              F2          -450.0    1868.742268      2318.742268   
4              F2          -450.0   21611.965380     22061.965380   
...           ...             ...            ...              ...   
1291           F2          -450.0  112341.304928    112791.304928   
1292           F2          -450.0  590805.599163    591255.599163   
1293           F2          -450.0  575018.988130    575468.988130   
1294           F2          -450.0  232243.327721    232693.327721   
1295           F2          -450.0  260139.410316    260589.410316   

      AverageFitness  AverageDiffFromOpt  TimePerRun  Generations  \
0       3.213329e+04        3.258329e+04    0.011128           50   
1       2.238459e+04        2.283459e+04    0.027806          150   
2       1.487255e+04        1.532255e+04    0.010348           50   
3       5.054819e+03        5.504819e+03    0.031108          150   
4       5.134586e+04        5.179586e+04    0.009959           50   
...              ...                 ...         ...          ...   
1291    1.461942e+05        1.466442e+05    0.321514          150   
1292    9.641332e+05        9.645832e+05    0.171331           50   
1293    1.290336e+06        1.290786e+06    0.513789          150   
1294    3.293918e+05        3.298418e+05    0.154155           50   
1295    2.829519e+05        2.834019e+05    0.470305          150   

      NoDimension  PopSize  TForTournament  CrossoverPerc  MutationRate  \
0              10        5               2            0.5           0.2   
1              10        5               2            0.5           0.2   
2              10        5               2            0.5           0.2   
3              10        5               2            0.5           0.2   
4              10        5               2            0.5           0.2   
...           ...      ...             ...            ...           ...   
1291           50       15               4            1.0           1.0   
1292           50       15               4            1.0           1.0   
1293           50       15               4            1.0           1.0   
1294           50       15               4            1.0           1.0   
1295           50       15               4            1.0           1.0   

      DecreasingMutationRate  Elite       Bounds  AverageOverRange  
0                       True      0  [-100, 100]                 5  
1                       True      0  [-100, 100]                 5  
2                       True      1  [-100, 100]                 5  
3                       True      1  [-100, 100]                 5  
4                      False      0  [-100, 100]                 5  
...                      ...    ...          ...               ...  
1291                    True      1  [-100, 100]                 5  
1292                   False      0  [-100, 100]                 5  
1293                   False      0  [-100, 100]                 5  
1294                   False      1  [-100, 100]                 5  
1295                   False      1  [-100, 100]                 5  

[1296 rows x 17 columns]

# F4
- Best Diff: 459.44 -> 150 Generation / 10 Dimension / 15 Pop Size / 4 T / 95% CR / 20% MR / DMR False / 1 Elite
- Avr. Diff: 1841.4 -> 150 Generation / 10 Dimension / 15 Pop Size / 2 T / 95%  CR / 20% MR / DMR True / 1 Elite

In [5]:
df.query("TestFunction == 'F4'")

TestFunction  OptimalMinimum    BestFitness  BestDiffFromOpt  \
1296           F4          -450.0   19648.328195     20098.328195   
1297           F4          -450.0   43889.355029     44339.355029   
1298           F4          -450.0    3226.647401      3676.647401   
1299           F4          -450.0    1952.865879      2402.865879   
1300           F4          -450.0   53798.825239     54248.825239   
...           ...             ...            ...              ...   
2587           F4          -450.0  161980.060540    162430.060540   
2588           F4          -450.0  404325.050245    404775.050245   
2589           F4          -450.0  474857.028769    475307.028769   
2590           F4          -450.0  367568.143286    368018.143286   
2591           F4          -450.0  318431.871318    318881.871318   

      AverageFitness  AverageDiffFromOpt  TimePerRun  Generations  \
1296    4.668801e+04        4.713801e+04    0.009338           50   
1297    5.329062e+04        5.374062e+04    0.028457          150   
1298    1.438899e+04        1.483899e+04    0.009620           50   
1299    6.795449e+03        7.245449e+03    0.032871          150   
1300    1.240192e+05        1.244692e+05    0.011265           50   
...              ...                 ...         ...          ...   
2587    2.346768e+05        2.351268e+05    0.328651          150   
2588    1.207540e+06        1.207990e+06    0.172086           50   
2589    1.385936e+06        1.386386e+06    0.524369          150   
2590    5.447924e+05        5.452424e+05    0.155512           50   
2591    4.299755e+05        4.304255e+05    0.470834          150   

      NoDimension  PopSize  TForTournament  CrossoverPerc  MutationRate  \
1296           10        5               2            0.5           0.2   
1297           10        5               2            0.5           0.2   
1298           10        5               2            0.5           0.2   
1299           10        5               2            0.5           0.2   
1300           10        5               2            0.5           0.2   
...           ...      ...             ...            ...           ...   
2587           50       15               4            1.0           1.0   
2588           50       15               4            1.0           1.0   
2589           50       15               4            1.0           1.0   
2590           50       15               4            1.0           1.0   
2591           50       15               4            1.0           1.0   

      DecreasingMutationRate  Elite       Bounds  AverageOverRange  
1296                    True      0  [-100, 100]                 5  
1297                    True      0  [-100, 100]                 5  
1298                    True      1  [-100, 100]                 5  
1299                    True      1  [-100, 100]                 5  
1300                   False      0  [-100, 100]                 5  
...                      ...    ...          ...               ...  
2587                    True      1  [-100, 100]                 5  
2588                   False      0  [-100, 100]                 5  
2589                   False      0  [-100, 100]                 5  
2590                   False      1  [-100, 100]                 5  
2591                   False      1  [-100, 100]                 5  

[1296 rows x 17 columns]

# F8
- Best Diff: 20.17 -> 150 Generation / 10 Dimension / 15 Pop Size / 4 T / 100% CR / 80% MR / DMR True / 1 Elite
- Avr. Diff: 20.47 -> Same

In [6]:
df.query("TestFunction == 'F8'")

TestFunction  OptimalMinimum  BestFitness  BestDiffFromOpt  \
2592           F8          -140.0  -118.828332        21.171668   
2593           F8          -140.0  -118.780497        21.219503   
2594           F8          -140.0  -119.053467        20.946533   
2595           F8          -140.0  -119.263744        20.736256   
2596           F8          -140.0  -118.832553        21.167447   
...           ...             ...          ...              ...   
3883           F8          -140.0  -118.752831        21.247169   
3884           F8          -140.0  -118.536631        21.463369   
3885           F8          -140.0  -118.496857        21.503143   
3886           F8          -140.0  -118.684637        21.315363   
3887           F8          -140.0  -118.711035        21.288965   

      AverageFitness  AverageDiffFromOpt  TimePerRun  Generations  \
2592     -118.601792           21.398208    0.011346           50   
2593     -118.632335           21.367665    0.034004          150   
2594     -118.934699           21.065301    0.015518           50   
2595     -119.165199           20.834801    0.046569          150   
2596     -118.571401           21.428599    0.017156           50   
...              ...                 ...         ...          ...   
3883     -118.683254           21.316746    0.383418          150   
3884     -118.443720           21.556280    0.194784           50   
3885     -118.436180           21.563820    0.586340          150   
3886     -118.661321           21.338679    0.175834           50   
3887     -118.683613           21.316387    0.532701          150   

      NoDimension  PopSize  TForTournament  CrossoverPerc  MutationRate  \
2592           10        5               2            0.5           0.2   
2593           10        5               2            0.5           0.2   
2594           10        5               2            0.5           0.2   
2595           10        5               2            0.5           0.2   
2596           10        5               2            0.5           0.2   
...           ...      ...             ...            ...           ...   
3883           50       15               4            1.0           1.0   
3884           50       15               4            1.0           1.0   
3885           50       15               4            1.0           1.0   
3886           50       15               4            1.0           1.0   
3887           50       15               4            1.0           1.0   

      DecreasingMutationRate  Elite     Bounds  AverageOverRange  
2592                    True      0  [-32, 32]                 5  
2593                    True      0  [-32, 32]                 5  
2594                    True      1  [-32, 32]                 5  
2595                    True      1  [-32, 32]                 5  
2596                   False      0  [-32, 32]                 5  
...                      ...    ...        ...               ...  
3883                    True      1  [-32, 32]                 5  
3884                   False      0  [-32, 32]                 5  
3885                   False      0  [-32, 32]                 5  
3886                   False      1  [-32, 32]                 5  
3887                   False      1  [-32, 32]                 5  

[1296 rows x 17 columns]

# F13
- Best Diff: 1.8799 -> 150 Generation / 10 Dimension / 15 Pop Size / 4 T / 95% CR / 20% MR / DMR True / 0 Elite
- Avr. Diff: 2.9072 -> 150 Generation / 10 Dimension / 10 Pop Size / 2 T / 95% CR / 20% MR / DMR True / 1 Elite


In [7]:
df.query("TestFunction == 'F13'")

TestFunction  OptimalMinimum  BestFitness  BestDiffFromOpt  \
0             F13          -130.0  -114.747967        15.252033   
1             F13          -130.0  -112.484813        17.515187   
2             F13          -130.0  -119.482405        10.517595   
3             F13          -130.0  -127.111143         2.888857   
4             F13          -130.0    68.014676       198.014676   
...           ...             ...          ...              ...   
1291          F13          -130.0    18.446939       148.446939   
1292          F13          -130.0  3078.646003      3208.646003   
1293          F13          -130.0  5464.382972      5594.382972   
1294          F13          -130.0  1636.088684      1766.088684   
1295          F13          -130.0  1598.512694      1728.512694   

      AverageFitness  AverageDiffFromOpt  TimePerRun  Generations  \
0         123.382008          253.382008    0.015493           50   
1         -81.314126           48.685874    0.047154          150   
2        -117.122011           12.877989    0.014168           50   
3        -122.551732            7.448268    0.043210          150   
4        1145.731194         1275.731194    0.015538           50   
...              ...                 ...         ...          ...   
1291       86.802009          216.802009    0.559082          150   
1292     5533.162406         5663.162406    0.246466           50   
1293     9402.327325         9532.327325    0.757659          150   
1294     2376.984034         2506.984034    0.231453           50   
1295     2085.922189         2215.922189    0.702414          150   

      NoDimension  PopSize  TForTournament  CrossoverPerc  MutationRate  \
0              10        5               2            0.5           0.2   
1              10        5               2            0.5           0.2   
2              10        5               2            0.5           0.2   
3              10        5               2            0.5           0.2   
4              10        5               2            0.5           0.2   
...           ...      ...             ...            ...           ...   
1291           50       15               4            1.0           1.0   
1292           50       15               4            1.0           1.0   
1293           50       15               4            1.0           1.0   
1294           50       15               4            1.0           1.0   
1295           50       15               4            1.0           1.0   

      DecreasingMutationRate  Elite   Bounds  AverageOverRange  
0                       True      0  [-3, 1]                 5  
1                       True      0  [-3, 1]                 5  
2                       True      1  [-3, 1]                 5  
3                       True      1  [-3, 1]                 5  
4                      False      0  [-3, 1]                 5  
...                      ...    ...      ...               ...  
1291                    True      1  [-3, 1]                 5  
1292                   False      0  [-3, 1]                 5  
1293                   False      0  [-3, 1]                 5  
1294                   False      1  [-3, 1]                 5  
1295                   False      1  [-3, 1]                 5  

[1296 rows x 17 columns]

# F17
- Best Diff: 219.97 -> 150 Generation / 10 Dimension / 10 Pop Size / 4 T / 50% CR / 80% MR / DMR True / 1 Elite
- Avr. Diff: 283.42 -> 150 Generation / 10 Dimension / 15 Pop Size / 2 T / 100% CR / 100% MR / DMR True / 1 Elite

In [8]:
df.query("TestFunction == 'F17'")

TestFunction  OptimalMinimum  BestFitness  BestDiffFromOpt  \
1296          F17           120.0   748.766137       628.766137   
1297          F17           120.0   591.077512       471.077512   
1298          F17           120.0   454.579857       334.579857   
1299          F17           120.0   403.606848       283.606848   
1300          F17           120.0  1369.054293      1249.054293   
...           ...             ...          ...              ...   
2587          F17           120.0   692.699934       572.699934   
2588          F17           120.0  1726.874128      1606.874128   
2589          F17           120.0  1809.362841      1689.362841   
2590          F17           120.0  1286.370225      1166.370225   
2591          F17           120.0  1332.571466      1212.571466   

      AverageFitness  AverageDiffFromOpt  TimePerRun  Generations  \
1296     1079.084525          959.084525    0.113645           50   
1297      951.976583          831.976583    0.335133          150   
1298      557.498490          437.498490    0.110451           50   
1299      495.791958          375.791958    0.332959          150   
1300     1429.394142         1309.394142    0.110137           50   
...              ...                 ...         ...          ...   
2587      825.222940          705.222940    4.463455          150   
2588     1875.775029         1755.775029    1.468171           50   
2589     1871.932725         1751.932725    4.641679          150   
2590     1564.226187         1444.226187    1.467616           50   
2591     1565.618000         1445.618000    4.645089          150   

      NoDimension  PopSize  TForTournament  CrossoverPerc  MutationRate  \
1296           10        5               2            0.5           0.2   
1297           10        5               2            0.5           0.2   
1298           10        5               2            0.5           0.2   
1299           10        5               2            0.5           0.2   
1300           10        5               2            0.5           0.2   
...           ...      ...             ...            ...           ...   
2587           50       15               4            1.0           1.0   
2588           50       15               4            1.0           1.0   
2589           50       15               4            1.0           1.0   
2590           50       15               4            1.0           1.0   
2591           50       15               4            1.0           1.0   

      DecreasingMutationRate  Elite   Bounds  AverageOverRange  
1296                    True      0  [-5, 5]                 5  
1297                    True      0  [-5, 5]                 5  
1298                    True      1  [-5, 5]                 5  
1299                    True      1  [-5, 5]                 5  
1300                   False      0  [-5, 5]                 5  
...                      ...    ...      ...               ...  
2587                    True      1  [-5, 5]                 5  
2588                   False      0  [-5, 5]                 5  
2589                   False      0  [-5, 5]                 5  
2590                   False      1  [-5, 5]                 5  
2591                   False      1  [-5, 5]                 5  

[1296 rows x 17 columns]

# F24
- Best Diff: 500.05 -> 150 Generation / 10 Dimension / 15 Pop Size / 2 T / 100% CR / 20% MR / DMR False / 1 Elite
- Avr. Diff: 836.26 -> 150 Generation / 10 Dimension / 10 Pop Size / 4 T / 100% CR / 80% MR / DMR True / 1 Elite

In [9]:
df.query("TestFunction == 'F24'")

TestFunction  OptimalMinimum  BestFitness  BestDiffFromOpt  \
0             F24           260.0  1581.379194      1321.379194   
1             F24           260.0  1716.297855      1456.297855   
2             F24           260.0  1257.358857       997.358857   
3             F24           260.0  1009.196942       749.196942   
4             F24           260.0  1706.802306      1446.802306   
...           ...             ...          ...              ...   
1291          F24           260.0  1604.662489      1344.662489   
1292          F24           260.0  2013.695892      1753.695892   
1293          F24           260.0  2022.287128      1762.287128   
1294          F24           260.0  1880.468349      1620.468349   
1295          F24           260.0  1851.241865      1591.241865   

      AverageFitness  AverageDiffFromOpt  TimePerRun  Generations  \
0        1772.169403         1512.169403    0.115887           50   
1        1859.331639         1599.331639    0.334574          150   
2        1496.301694         1236.301694    0.095941           50   
3        1260.036581         1000.036581    0.290794          150   
4        1905.026391         1645.026391    0.097299           50   
...              ...                 ...         ...          ...   
1291     1633.768568         1373.768568    3.663990          150   
1292     2032.972450         1772.972450    1.252370           50   
1293     2078.779992         1818.779992    3.872165          150   
1294     1907.107345         1647.107345    1.220936           50   
1295     1866.000102         1606.000102    3.771049          150   

      NoDimension  PopSize  TForTournament  CrossoverPerc  MutationRate  \
0              10        5               2            0.5           0.2   
1              10        5               2            0.5           0.2   
2              10        5               2            0.5           0.2   
3              10        5               2            0.5           0.2   
4              10        5               2            0.5           0.2   
...           ...      ...             ...            ...           ...   
1291           50       15               4            1.0           1.0   
1292           50       15               4            1.0           1.0   
1293           50       15               4            1.0           1.0   
1294           50       15               4            1.0           1.0   
1295           50       15               4            1.0           1.0   

      DecreasingMutationRate  Elite   Bounds  AverageOverRange  
0                       True      0  [-5, 5]                 5  
1                       True      0  [-5, 5]                 5  
2                       True      1  [-5, 5]                 5  
3                       True      1  [-5, 5]                 5  
4                      False      0  [-5, 5]                 5  
...                      ...    ...      ...               ...  
1291                    True      1  [-5, 5]                 5  
1292                   False      0  [-5, 5]                 5  
1293                   False      0  [-5, 5]                 5  
1294                   False      1  [-5, 5]                 5  
1295                   False      1  [-5, 5]                 5  

[1296 rows x 17 columns]

# Conclusion:

Frequency of configuration among best difference and average difference from optimal:
- Generation:
    - 50 -> 0
    - 150 -> 12
- Dimension:
    - 10 -> 12
    - 30 -> 0
    - 50 -> 0
- Population Size:
    - 5 -> 0
    - 10 -> 3
    - 15 -> 9
- Tournament Size:
    - 2 -> 5
    - 4 -> 7
- Crossover Rate:
    - 50% -> 1
    - 95% -> 5
    - 100% -> 6
- Mutation Rate:
    - 50% -> 1
    - 95% -> 5
    - 100% -> 6
- Decreasing Mutation Rate:
    - True -> 10
    - False -> 2
- Elites:
    - 0 -> 1
    - 1 -> 11

F8 and F13 performed the best. F13 error was the lowest with the best being 1.44% away from the optimal. F8's best fitness was 15.51% away from the optimal. This however is a significant error and can be approved. As such for the other

# 2nd Iteration of Configurations

In [10]:
benchmarks_and_bounds = [optproblems.F2, [-100,100], "F2", optproblems.F4, [-100,100], "F4", optproblems.F8, [-32,32], "F8", optproblems.F13, [-3,1], "F13", optproblems.F17, [-5,5], "F17", optproblems.F24, [-5,5], "F24"]
no_dimensions = [10]
num_generations = [250, 500, 1000]
pop_sizes = [10, 15, 20]
tournament_t = [2, 4]
crossover_percentage = [0.95, 1]
mutation_rates = [0.2, 0.8]
decreasing_mutation_rate = [True]
elites = [0, 1]

average_over_range = 5

In [ ]:
run_configuration_suite = True
if run_configuration_suite:

    list_for_df = []
    for i in range(0, len(benchmarks_and_bounds), 3):
        number_of_functions = (len(benchmarks_and_bounds)/3)
        print(f"Progress: {i/3} / {number_of_functions}")
        bound = benchmarks_and_bounds[i+1]
        benchmark_name = benchmarks_and_bounds[i+2]
        for no_dimension in no_dimensions:
            print(f"\tDimension: {no_dimension}")
            benchmark = benchmarks_and_bounds[i](no_dimension)
            optimal_solution = benchmark.get_optimal_solutions()
            benchmark.evaluate(optimal_solution[0])
            optimal_minimum = optimal_solution[0].objective_values
            for pop_size in pop_sizes:
                print(f"\t\tPopulation: {pop_size}")
                for t in tournament_t:
                    for crossover in crossover_percentage:
                        for mutation in mutation_rates:
                            for dmr in decreasing_mutation_rate:
                                for elite in elites:
                                    for generations in num_generations:
                                        average_fitness = 0
                                        average_difference_from_optimal = 0

                                        best_fitness = 0
                                        best_time = 0
                                        best_difference_from_optimal = 0

                                        for j in range(average_over_range):
                                            algorithm = GA(benchmark, no_dimension, bound, pop_size, num_generations=generations, t=t, crossover_percentage=crossover, mutation_rate=mutation, decreasing_mutation_rate=dmr, elite=elite)
                                            result, time = algorithm.run()
                                            fitness = result.fitness
                                            diff_from_optimal = fitness - optimal_minimum

                                            average_fitness += 1/average_over_range * fitness
                                            average_difference_from_optimal += 1/average_over_range * diff_from_optimal

                                            if j == 0:
                                                best_fitness = fitness
                                                best_difference_from_optimal = diff_from_optimal
                                                best_time = time
                                                continue

                                            if fitness < best_fitness:
                                                best_fitness = fitness

                                            if diff_from_optimal < best_difference_from_optimal:
                                                best_difference_from_optimal = diff_from_optimal

                                            if time < best_time:
                                                best_time = time

                                        result_and_configuration = [benchmark_name, optimal_minimum, best_fitness, best_difference_from_optimal, average_fitness, average_difference_from_optimal, best_time, generations, no_dimension, pop_size, t, crossover, mutation, dmr, elite, bound, average_over_range]
                                        list_for_df.append(result_and_configuration)

    labels = ["TestFunction", "OptimalMinimum", "BestFitness", "BestDiffFromOpt", "AverageFitness", "AverageDiffFromOpt", "TimePerRun", "Generations", "NoDimension", "PopSize", "TForTournament", "CrossoverPerc", "MutationRate", "DecreasingMutationRate", "Elite", "Bounds", "AverageOverRange"]
    df_II = pandas.DataFrame(data=list_for_df, columns=labels)
    df_II.to_pickle("GA_df_specific.pkl")
else:
    df_II = pd.read_pickle("GA_df_specific.pkl")